File: data/src/main/java/com/ykb/app/cryptotrader/data/model/BaseEntity.java
package com.ykb.app.cryptotrader.data.model;

import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.Temporal;
import jakarta.persistence.TemporalType;
import lombok.Data;

import java.util.Date;

@MappedSuperclass
@Data
abstract class BaseEntity {

    @Column(name = "CREATION_TIME")
    @Temporal(TemporalType.TIMESTAMP)
    private Date creationTime = new Date();

    @Column(name = "UPDATE_TIME")
    @Temporal(TemporalType.TIMESTAMP)
    private Date updateTime = null;

    @Column(name = "INACTIVATION_TIME")
    @Temporal(TemporalType.TIMESTAMP)
    private Date inactivationTime = null;

    public void update() {
        updateTime = new Date();
    }

    public void inactivate() {
        updateTime = new Date();
        inactivationTime = updateTime;
    }

}

File: library/data/data-base/src/main/kotlin/com/ykb/app/cryptotrader/data/base/entity/BaseEntity.kt
package com.ykb.app.cryptotrader.data.base.entity

import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.Temporal
import jakarta.persistence.TemporalType
import java.util.Date

@MappedSuperclass
abstract class BaseEntity {

    @Column(name = "CREATION_TIME")
    @Temporal(TemporalType.TIMESTAMP)
    private val creationTime = Date()

    @Column(name = "UPDATE_TIME")
    @Temporal(TemporalType.TIMESTAMP)
    private var updateTime: Date? = null

    @Column(name = "INACTIVATION_TIME")
    @Temporal(TemporalType.TIMESTAMP)
    private var inactivationTime: Date? = null

    fun getCreationTime(): Date {
        return creationTime
    }

    fun getUpdateTime(): Date? {
        return updateTime
    }

    fun update() {
        updateTime = Date()
    }

    fun getInactivationTime(): Date? {
        return inactivationTime
    }

    fun inactivate() {
        updateTime = Date()
        inactivationTime = updateTime
    }

}